### PR TITLE
fix: forget password and sign up router

### PR DESCRIPTION
**Pull request by imchell**
Fix: #1181 

page router

### CODE REVIEW

**Comment by hsluoyz** [comment]
@leo220yuyaodog plz review

File: web/src/Setting.js [hunk 1]
```diff
@@ -797,6 +818,22 @@ export function goToForget(ths, application) {
...
+    return "/";
+  }
+
+  if (authConfig.appName === application.name) {
+    return "/forget";
+  } else {
+    if (application.forgetUrl === "") {
+      return `/forget/${application.name}`;
+    } else {
+      return "application.forgetUrl";
```
Comment (hsluoyz) [inline web/src/Setting.js:832]: `application.forgetUrl` can be other domain's URL

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by leo220yuyaodog** [comment]
@imchell Actually, now from the login page fo to the signup and forget page is the SPA router. It can not invoke the page refresh.  From the signup page to login page will refresh the browser. 

When switch the pages will appears a short white pages, because react renders background image. The backgroud image url is from the react state `application`. And the `LoginPage`, `SignupPage` have their owe state  `application`, so react renders background iamge again, cause the problem. 

I have an idea that extract the background div to the father component. Then when switch the pages, the background will not render again.  Do you hava other good implenmentation metheds?

**Comment by imchell** [comment]
@leo220yuyaodog Thanks for reviewing. Yes, the background page reloads while switching between `LoginPage` and `SignupPage`. I'm not a react-router pro but I don't know why background being extracted to the parent component helps. It is foreseeable that the extraction would lead to the same background image of all pages in `<Switch>`, which is not the intended effect, thus needing some work on decomposing elements in `<Switch>`.

**Comment by leo220yuyaodog** [comment]
> @leo220yuyaodog Thanks for reviewing. Yes, the background page reloads while switching between `LoginPage` and `SignupPage`. I'm not a react-router pro but I don't know why background being extracted to the parent component helps. It is foreseeable that the extraction would lead to the same background image of all pages in `<Switch>`, which is not the intended effect.

I don't think about it enough. Fixing the front-end routing bug is probably the best thing to do right now, if it can resolve the problem.

File: web/src/Setting.js [hunk 2]
```diff
@@ -781,6 +782,27 @@ export function goToSignup(ths, application) {
   }
 }
 
+export function renderDynamicSignUpLink(application, children) {
```
Comment (hsluoyz) [inline web/src/Setting.js:785]: All `children` to `text`

File: web/src/Setting.js [hunk 3]
```diff
@@ -781,6 +782,27 @@ export function goToSignup(ths, application) {
...
+  if (!application.enablePassword && window.location.pathname.includes("/login/oauth/authorize")) {
+    const link = window.location.href.replace("/login/oauth/authorize", "/auto-signup/oauth/authorize");
+    return <a style={{float: "right"}} href={link} target="_blank" rel="noopener noreferrer">{children}</a>;
+  }
+
+  if (authConfig.appName === application.name) {
+    return <Link to={"/signup"}>{children}</Link>;
+  } else {
+    if (application.signupUrl === "") {
+      return <Link to={`/signup/${application.name}`}>{children}</Link>;
```
Comment (hsluoyz) [inline web/src/Setting.js:799]: Extract a function, use `url` as arg

File: web/src/Setting.js (hunk 2 above)
Comment (hsluoyz) [inline web/src/Setting.js:785]: renderSignupLink

File: web/src/Setting.js [hunk 4]
```diff
@@ -797,6 +819,22 @@ export function goToForget(ths, application) {
   }
 }
 
+export function renderDynamicForgetLink(application, children) {
```
Comment (hsluoyz) [inline web/src/Setting.js:822]: renderForgetLink

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/auth/LoginPage.js [hunk 5]
```diff
@@ -473,12 +471,9 @@ class LoginPage extends React.Component {
               !application.enableSignUp ? null : (
                 <>
                   {i18next.t("login:No account?")}&nbsp;
-                  <a onClick={() => {
-                    sessionStorage.setItem("signinUrl", window.location.href);
```
Comment (hsluoyz) [inline web/src/auth/LoginPage.js:477]: `sessionStorage.setItem("signinUrl", window.location.href);` is missing

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**